src/transformers/models/vit/convert_vit_timm_to_pytorch.py

```diff
@@ -140,60 +140,39 @@ def convert_vit_checkpoint(vit_name, pytorch_dump_folder_path):
     # define default ViT configuration
     config = ViTConfig()
     base_model = False
-    # dataset (ImageNet-21k only or also fine-tuned on ImageNet 2012), patch_size and image_size
-    if vit_name[-5:] == "in21k":
-        base_model = True
-        config.patch_size = int(vit_name[-12:-10])
-        config.image_size = int(vit_name[-9:-6])
-    else:
-        config.num_labels = 1000
+
+    # load original model from timm
+    timm_model = timm.create_model(vit_name, pretrained=True)
+    timm_model.eval()
+
+    # get patch size and image size from the patch embedding submodule
+    config.patch_size = timm_model.patch_embed.patch_size[0]
+    config.image_size = timm_model.patch_embed.img_size[0]
+
+    # retrieve architecture-specific parameters from the timm model
+    config.hidden_size = timm_model.embed_dim
+    config.intermediate_size = timm_model.blocks[0].mlp.fc1.out_features
+    config.num_hidden_layers = len(timm_model.blocks)
+    config.num_attention_heads = timm_model.blocks[0].attn.num_heads
+
+    # check whether the model has a classification head or not
+    if timm_model.num_classes != 0:
+        config.num_labels = timm_model.num_classes
```

Comment:
So, this block here assumes in1k is the only situation when num_classes != 0 which is certainly not true these days, it was back when vit was first released, only models with valid classifiers were 1k (21k were all zero'd). 

Now timm has following layouts for vits that I can see:
* 1000 - Imagenet-1k 
* 21843 - Google's ImageNet-21k layout, from How to Train Your ViT paper (differs from other standard 21/22k layouts with 21841 classes)
* 11821 - my 'ImageNet-12k' layout
* A few models from CLIP weights that have a final embedding projection as num_classes (512, 768, etc), should not map them to classes

So, to be correct and get the mapping, something like that needs to be used https://github.com/huggingface/api-inference-community/blob/main/docker_images/timm/app/pipelines/image_classification.py#L31-L33 

However, I will point out this reveals the flaw in the id2label/label2id scheme, it's not adequate to cover situation like ImageNet-21k when being used as a 2-way mapping that assumes uniquenes. https://huggingface.co/datasets/imagenet-1k/discussions/1#651c5d13aa89f7554c3e26ca

Reply:
I have tried to fix this by using the 'infer_imagenet_subset' function from timm.data. I have used the index_to_label_name() method from ImageNetInfo to get the label names. Let me know if this is not how it should be done